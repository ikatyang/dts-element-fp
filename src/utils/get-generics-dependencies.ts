import * as dts from 'dts-element';
import { has } from './has';

import { IGenericDeclaration } from 'dts-element';

export const get_generics_dependencies = (
  generics: dts.IGenericDeclaration[],
) =>
  generics.map(generic => [
    generic,
    ...generics.filter(current_generic =>
      has(generic, {
        kind: dts.ElementKind.GeneralType,
        name: current_generic.name,
      }),
    ),
  ]);
