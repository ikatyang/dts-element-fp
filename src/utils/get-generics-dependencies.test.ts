import * as dts from 'dts-element';
import { get_generics_dependencies } from './get-generics-dependencies';

test_case(test, '<A, B extends A, C extends A, D, E>');
test_case(test, '<A, B extends A, C extends B, D, E>');

function test_case(tester: typeof test, generics: string) {
  tester(generics, () => {
    expect(
      emit_generics_dependencies(parse_generices(generics)),
    ).toMatchSnapshot();
  });
}

function parse_generices(str: string) {
  const type = dts.parse_type(`${str}() => any`);
  return (type as dts.IFunctionType).generics!;
}

function emit_generics_dependencies(generics: dts.IGenericDeclaration[]) {
  const generics_dependencies = get_generics_dependencies(generics);
  return generics.reduce(
    (current, generic, index) => ({
      ...current,
      [generic.name]: generics_dependencies[index].map(x => x.name),
    }),
    {},
  );
}
