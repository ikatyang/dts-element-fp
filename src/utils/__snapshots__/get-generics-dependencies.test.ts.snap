// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<A, B extends A, C extends A, D, E> 1`] = `
Object {
  "A": Array [
    "A",
  ],
  "B": Array [
    "B",
    "A",
  ],
  "C": Array [
    "C",
    "A",
  ],
  "D": Array [
    "D",
  ],
  "E": Array [
    "E",
  ],
}
`;

exports[`<A, B extends A, C extends B, D, E> 1`] = `
Object {
  "A": Array [
    "A",
  ],
  "B": Array [
    "B",
    "A",
  ],
  "C": Array [
    "C",
    "B",
    "A",
  ],
  "D": Array [
    "D",
  ],
  "E": Array [
    "E",
  ],
}
`;
